### PR TITLE
implement traffic switching based on internal code

### PR DESCRIPTION
**Pull request by szuecs**
The below example, shows a possible eskip document used for green-blue
deployments of APIS, which usually don't require stickiness:

    // hit by 10% percent chance
    v2:
        Traffic(.1) ->
        "https://api-test-green";

    // hit by remaining chance
    v1:
        "https://api-test-blue";

stickyness example:

    // hit by 5% percent chance
    cartTest:
        Traffic(.05, "cart-test", "test") && Path("/cart") ->
        responseCookie("cart-test", "test") ->
        "https://cart-test"\;

    // hit by remaining chance
    cart:
        Path("/cart") ->
        responseCookie("cart-test", "default") ->
        "https://cart"\;

### CODE REVIEW

**Comment by lmineiro** [comment]
I think for APIs the predicate could be simplified and require only the first arg - the percentage

**Comment by szuecs** [comment]
@lmineiro good point, fixed in the last commit

File: predicates/traffic/traffic.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
+/*
+Package traffic implements a custom predicate to control
+traffic by setting rates for certain routes.
```
Comment (lmineiro) [inline predicates/traffic/traffic.go]: "Package traffic implements a predicate to control the matching probability for a given route by setting its weight"

File: predicates/traffic/traffic.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+The chance for hitting a route is defined by the mandatory first
+parameter, that must be a number between 0 and 1.
+
+The optional second argument is used to specify the cookie
+name for the traffic group.
```
Comment (lmineiro) [inline predicates/traffic/traffic.go]: I'd also add a short explanation. "Stickiness allows all subsequent requests form the same client to match the same route"

File: predicates/traffic/traffic.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+func (s *spec) Create(args []interface{}) (routing.Predicate, error) {
+	if !(len(args) == 1 || len(args) == 3) {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	p := &predicate{}
+
+	if c, ok := args[0].(float64); ok {
+		p.chance = c
```
Comment (lmineiro) [inline predicates/traffic/traffic.go]: Shouldn't we validate that `0.0 <= c < 1.0` ?

File: predicates/traffic/traffic_test.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+func TestMatch(t *testing.T) {
+	for _, ti := range []struct {
+		msg   string
+		p     predicate
+		r     http.Request
+		match bool
+	}{{
+		"not sticky, no match",
```
Comment (lmineiro) [inline predicates/traffic/traffic_test.go]: We could add a controlled randomness test by setting the seed

File: predicates/traffic/traffic.go (hunk 3 above)
Comment (szuecs) [inline predicates/traffic/traffic.go]: you mean ` 0.0 <= c && c <= 1.0`, right (both <=)?

File: predicates/traffic/traffic_test.go (hunk 4 above)
Comment (szuecs) [inline predicates/traffic/traffic_test.go]: I am not sure if I get this point. What does this help to test?

**Comment by lmineiro** [comment]
👍

**Comment by szuecs** [comment]
👍

**Pull request merged**